browser/brave_vpn/dns/brave_vpn_dns_observer_service.cc

```diff
@@ -0,0 +1,277 @@
+    net::DnsOverHttpsConfig doh_config({entry->doh_server_config});
+    return doh_config.ToString();
+  }
+  NOTREACHED() << "Should not be reached as we expect " << kBraveVpnDnsProvider
+               << " is available in the default list.";
+  return std::string("1.1.1.1");
+}
+
+std::string GetDoHServers(const std::string* user_servers) {
+  return user_servers && !user_servers->empty()
+             ? *user_servers
+             : GetFilteredProvidersForCountry();
+}
+
+gfx::NativeWindow GetAnchorBrowserWindow() {
+  auto* browser = chrome::FindLastActive();
+  return browser ? browser->window()->GetNativeWindow()
+                 : gfx::kNullNativeWindow;
+}
+
+}  // namespace
+
+BraveVpnDnsObserverService::BraveVpnDnsObserverService(
+    PrefService* local_state,
+    PrefService* profile_prefs,
+    DnsPolicyReaderCallback callback)
+    : policy_reader_(std::move(callback)),
+      local_state_(local_state),
+      profile_prefs_(profile_prefs) {
+  DCHECK(profile_prefs_);
+  DCHECK(local_state_);
+
+  net::NetworkChangeNotifier::AddDNSObserver(this);
+  pref_change_registrar_.Init(local_state);
+  pref_change_registrar_.Add(
+      ::prefs::kDnsOverHttpsMode,
+      base::BindRepeating(&BraveVpnDnsObserverService::OnDNSPrefChanged,
+                          weak_ptr_factory_.GetWeakPtr()));
+  pref_change_registrar_.Add(
+      ::prefs::kDnsOverHttpsTemplates,
+      base::BindRepeating(&BraveVpnDnsObserverService::OnDNSPrefChanged,
+                          weak_ptr_factory_.GetWeakPtr()));
+}
+
+BraveVpnDnsObserverService::~BraveVpnDnsObserverService() {
+  net::NetworkChangeNotifier::RemoveDNSObserver(this);
+}
+
+void BraveVpnDnsObserverService::OnDNSChanged() {
+  // Re-create service for reading config once becasue the service keeps
+  // the callback and cannot read multiple times.
+  dns_config_service_ = net::DnsConfigService::CreateSystemService();
+  dns_config_service_->ReadConfig(
+      base::BindRepeating(&BraveVpnDnsObserverService::OnSystemDNSConfigChanged,
+                          weak_ptr_factory_.GetWeakPtr()));
+}
+
+void BraveVpnDnsObserverService::OnSystemDNSConfigChanged(
+    const net::DnsConfig& config) {
+  system_dns_config_ = config;
+  dns_config_service_.reset();
+}
+
+bool BraveVpnDnsObserverService::IsDNSSecure(
+    SecureDnsConfig browser_dns_config) const {
+  bool secure = (SecureDnsConfig::ModeToString(browser_dns_config.mode()) ==
+                 SecureDnsConfig::kModeSecure);
+  bool is_valid_automatic_mode =
+      (SecureDnsConfig::ModeToString(browser_dns_config.mode()) ==
+       SecureDnsConfig::kModeAutomatic) &&
+      IsValidDoHTemplates(browser_dns_config.doh_servers().ToString());
+  bool is_system_doh_enabled =
+      (!system_dns_config_.doh_config.servers().empty() &&
+       system_dns_config_.allow_dns_over_https_upgrade);
+  return secure || is_valid_automatic_mode || is_system_doh_enabled;
+}
+
+bool BraveVpnDnsObserverService::ShouldAllowExternalChanges() const {
```

Comment:
should allow what changes?

Reply:
I think this is just for unit testing purposes - otherwise it just returns the result of the message box

Reply:
The dialog itself says:
```
Turning off secure DNS or using default servers on Windows may reveal the names of the sites you visit to your Internet service provider. Brave does not recommend doing this. Do you want to do it anyways?
```

Reply:
if user has DoH enabled we do not overwrite the settings and enable VPN, after this user can try to disable it, or DoH could be disabled by third party tool/extension/program(in case of the system DoH). In this cases we notify user to ask if we have to restore old settings.

Reply:
comments are needed in the header file explaining what these methods do

Reply:
also I think `ShouldAllowDohOverride` would be better

Reply:
done